server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessageVaultRestoreTask.java

```diff
@@ -0,0 +1,174 @@
+import org.apache.james.task.Task;
+import org.apache.james.task.TaskExecutionDetails;
+import org.apache.james.vault.DeletedMessage;
+import org.apache.james.vault.DeletedMessageVault;
+import org.apache.james.vault.Query;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+class DeletedMessageVaultRestoreTask implements Task {
+
+    public static class AdditionalInformation implements TaskExecutionDetails.AdditionalInformation {
+        private final User user;
+        private final long successfulRestoreCount;
+        private final long errorRestoreCount;
+        private final long failRemovingCount;
+
+        AdditionalInformation(User user, long successfulRestoreCount, long errorRestoreCount, long failRemovingCount) {
+            this.user = user;
+            this.successfulRestoreCount = successfulRestoreCount;
+            this.errorRestoreCount = errorRestoreCount;
+            this.failRemovingCount = failRemovingCount;
+        }
+
+        public long getSuccessfulRestoreCount() {
+            return successfulRestoreCount;
+        }
+
+        public long getErrorRestoreCount() {
+            return errorRestoreCount;
+        }
+
+        public long getFailRemovingCount() {
+            return failRemovingCount;
+        }
+
+        public String getUser() {
+            return user.asString();
+        }
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DeletedMessageVaultRestoreTask.class);
+
+    static final String TYPE = "deletedMessages/restore";
+    static final String RESTORE_MAILBOX_NAME = "restored messages";
+
+    private final DeletedMessageVault deletedMessageVault;
+    private final User userToRestore;
+    private final MailboxManager mailboxManager;
+    private final MailboxPath restoreMailbox;
+
+    private final AtomicLong successfulRestoreCount;
+    private final AtomicLong errorRestoreCount;
+    private final AtomicLong failRemovingCount;
+
+    DeletedMessageVaultRestoreTask(DeletedMessageVault deletedMessageVault, User userToRestore, MailboxManager mailboxManager) {
+        this.deletedMessageVault = deletedMessageVault;
+        this.userToRestore = userToRestore;
+        this.mailboxManager = mailboxManager;
+        this.restoreMailbox = MailboxPath.forUser(userToRestore.asString(), RESTORE_MAILBOX_NAME);
+        this.successfulRestoreCount = new AtomicLong();
+        this.errorRestoreCount = new AtomicLong();
+        this.failRemovingCount = new AtomicLong();
+    }
+
+    @Override
+    public Result run() {
+        try {
+            MailboxSession session = mailboxManager.createSystemSession(userToRestore.asString());
+            MessageManager restoreMailboxManager = restoreMailboxManager(session);
+
+            return Flux.from(deletedMessageVault.search(userToRestore, Query.ALL)).toStream()
```

Comment:
why do you switch to Stream?

Reply:
my first implementation is using flux pipleline, but I got a weird exception, it takes me hours but still can not find the cause. I switched to use a normal stream stuff and no more error

Reply:
found the cause. It is the `MemoryDeletedMessagesVault` HashTable concurrency issue. I will update the fix then turn to use reactor pipeline.